File: internal/transport/http/masterdata.go
package http

import (
	"github.com/go-chi/chi/v5"

	"github.com/fadelmajid/tera/internal/domain/money"
	"github.com/fadelmajid/tera/internal/service"
	"github.com/fadelmajid/tera/internal/store/gen"
)

// --- wire shapes ------------------------------------------------------------
//
// Deliberately not the sqlc row types. Those carry storage spellings
// (SalePriceIdr) and storage booleans (int64), and money must cross the wire as
// a branded integer the TypeScript side can trust (SPEC §1, INV-1).

type entityDTO struct {
	ID                 string  `json:"id"`
	Code               string  `json:"code"`
	Name               string  `json:"name"`
	IsPKP              bool    `json:"is_pkp"`
	NPWP               *string `json:"npwp"`
	Timezone           string  `json:"timezone"`
	BookYearStartMonth int64   `json:"book_year_start_month"`
	IsActive           bool    `json:"is_active"`
}

func toEntityDTO(e gen.LegalEntity) entityDTO {
	return entityDTO{
		ID: e.ID, Code: e.Code, Name: e.Name, IsPKP: e.IsPkp == 1, NPWP: e.Npwp,
		Timezone: e.Timezone, BookYearStartMonth: e.BookYearStartMonth, IsActive: e.IsActive == 1,
	}
}

type ownerDTO struct {
	ID       string  `json:"id"`
	Code     string  `json:"code"`
	Name     string  `json:"name"`
	Note     *string `json:"note"`
	IsActive bool    `json:"is_active"`
}

func toOwnerDTO(o gen.Owner) ownerDTO {
	return ownerDTO{ID: o.ID, Code: o.Code, Name: o.Name, Note: o.Note, IsActive: o.IsActive == 1}
}

type productDTO struct {
	ID       string  `json:"id"`
	Code     string  `json:"code"`
	Barcode  *string `json:"barcode"`
	Name     string  `json:"name"`
	Unit     string  `json:"unit"`
	Category *string `json:"category"`
	// Null is the company bucket, and it is reported as its own line on the
	// margin report (R2.2). Explicitly null rather than omitted, so the
	// distinction survives the wire.
	OwnerID      *string   `json:"owner_id"`
	SalePriceIDR money.IDR `json:"sale_price_idr"`
	IsActive     bool      `json:"is_active"`
}

func toProductDTO(p gen.Product) productDTO {
	return productDTO{
		ID: p.ID, Code: p.Code, Barcode: p.Barcode, Name: p.Name, Unit: p.Unit,
		Category: p.Category, OwnerID: p.OwnerID,
		SalePriceIDR: money.IDR(p.SalePriceIdr), IsActive: p.IsActive == 1,
	}
}

type supplierDTO struct {
	ID           string  `json:"id"`
	Code         string  `json:"code"`
	Name         string  `json:"name"`
	NPWP         *string `json:"npwp"`
	Address      *string `json:"address"`
	Phone        *string `json:"phone"`
	IssuesFaktur bool    `json:"issues_faktur"`
	IsActive     bool    `json:"is_active"`
}

func toSupplierDTO(s gen.Supplier) supplierDTO {
	return supplierDTO{
		ID: s.ID, Code: s.Code, Name: s.Name, NPWP: s.Npwp, Address: s.Address,
		Phone: s.Phone, IssuesFaktur: s.IssuesFaktur == 1, IsActive: s.IsActive == 1,
	}
}

type customerDTO struct {
	ID       string  `json:"id"`
	Code     string  `json:"code"`
	Name     string  `json:"name"`
	NPWP     *string `json:"npwp"`
	NIK      *string `json:"nik"`
	Address  *string `json:"address"`
	Phone    *string `json:"phone"`
	IsActive bool    `json:"is_active"`
}

func toCustomerDTO(c gen.Customer) customerDTO {
	return customerDTO{
		ID: c.ID, Code: c.Code, Name: c.Name, NPWP: c.Npwp, NIK: c.Nik,
		Address: c.Address, Phone: c.Phone, IsActive: c.IsActive == 1,
	}
}

// --- routes -----------------------------------------------------------------

func mountMasterData(r chi.Router, cfg Config) {
	md, auth := cfg.Master, cfg.Auth
	if md == nil {
		return
	}

	// Listing companies needs only a session: the browser has to know which
	// companies exist before it can name one in X-Entity-Id.
	r.Group(func(r chi.Router) {
		r.Use(requireAuth)
		r.Get("/entities", handleListEntities(md))
		r.Post("/setup/entity", handleSetupFirstEntity(md, auth))
	})

	// Everything else is scoped to a company and gated on the role held there.
	r.Group(func(r chi.Router) {
		r.Use(requireEntityRole(service.Role.CanManageMasterData,
			"hanya pemilik dan manajer yang dapat mengubah data master"))

		r.Post("/owners", handleCreate[ownerRequest](md.CreateOwner, toOwnerDTO))
		r.Put("/owners/{id}", handleUpdate[ownerRequest](md.UpdateOwner, toOwnerDTO))
		r.Post("/products", handleCreate[productRequest](md.CreateProduct, toProductDTO))
		r.Put("/products/{id}", handleUpdate[productRequest](md.UpdateProduct, toProductDTO))
		r.Post("/suppliers", handleCreate[supplierRequest](md.CreateSupplier, toSupplierDTO))
		r.Put("/suppliers/{id}", handleUpdate[supplierRequest](md.UpdateSupplier, toSupplierDTO))
		r.Post("/customers", handleCreate[customerRequest](md.CreateCustomer, toCustomerDTO))
		r.Put("/customers/{id}", handleUpdate[customerRequest](md.UpdateCustomer, toCustomerDTO))
	})

	// Reading master data needs a role in the company, but any of the three:
	// a cashier has to see the products they are selling.
	r.Group(func(r chi.Router) {
		r.Use(requireEntityRole(anyRole, "tidak punya akses"))

		r.Get("/owners", handleList(md.ListOwners, toOwnerDTO))
		r.Get("/products", handleList(md.ListProducts, toProductDTO))
		r.Get("/suppliers", handleList(md.ListSuppliers, toSupplierDTO))
		r.Get("/customers", handleList(md.ListCustomers, toCustomerDTO))
	})
}

// anyRole admits all three roles. Spelled out rather than passed as a literal
// so the intent is greppable next to the capability predicates.
func anyRole(r service.Role) bool { return r.Valid() }

File: web/src/api/masterdata.ts
import { api } from './client'
import type { IDR } from '../money'

/**
 * A company. Two exist: one PKP, one not, and they behave differently on every
 * sale and purchase (R1.3, SPEC §2.3).
 */
export interface Entity {
  id: string
  code: string
  name: string
  is_pkp: boolean
  npwp: string | null
  timezone: string
  book_year_start_month: number
  is_active: boolean
}

/** A family member products are attributed to. An attribution tag, not legal ownership (R2). */
export interface Owner {
  id: string
  code: string
  name: string
  note: string | null
  is_active: boolean
}

export interface Product {
  id: string
  code: string
  barcode: string | null
  name: string
  unit: string
  category: string | null
  /** null is the company bucket — its own line on the margin report (R2.2). */
  owner_id: string | null
  /**
   * Whole rupiah. The server stores it as int64 in a STRICT column, so the
   * value arriving here is always an integer (INV-1); user input goes through
   * parseIDR, which refuses anything fractional.
   */
  sale_price_idr: IDR
  is_active: boolean
}

export interface Supplier {
  id: string
  code: string
  name: string
  npwp: string | null
  address: string | null
  phone: string | null
  /**
   * Whether this supplier NORMALLY issues a faktur pajak. It defaults the
   * purchasing screen and lets suppliers be compared on true cost (R10.6). It
   * is never the truth for a given purchase — that is recorded per delivery and
   * decides the stock layer's cost basis (INV-9).
   */
  issues_faktur: boolean
  is_active: boolean
}

export interface Customer {
  id: string
  code: string
  name: string
  npwp: string | null
  nik: string | null
  address: string | null
  phone: string | null
  is_active: boolean
}

export const listEntities = () => api<Entity[]>('/entities')

export const setupFirstEntity = (body: {
  code: string
  name: string
  is_pkp: boolean
  npwp?: string
  timezone?: string
  book_year_start_month?: number
}) => api<Entity>('/setup/entity', { method: 'POST', body })

/**
 * Adds another company. Owner only.
 *
 * setupFirstEntity handles exactly the first one; every company after it comes
 * through here. The two companies are what make transfers, per-entity PKP
 * behaviour, and the omzet clock mean anything (R1.1, R1.2).
 */
export const createEntity = (
  entityId: string,
  body: {
    code: string
    name: string
    is_pkp: boolean
    npwp?: string
    timezone?: string
    book_year_start_month?: number
  },
) => api<Entity>('/entities', { method: 'POST', entityId, body })

const scoped =
  <T>(path: string) =>
  (entityId: string) =>
    api<T[]>(path, { entityId })

const creator =
  <T, B>(path: string) =>
  (entityId: string, body: B) =>
    api<T>(path, { method: 'POST', body, entityId })

const updater =
  <T, B>(path: string) =>
  (entityId: string, id: string, body: B) =>
    api<T>(`${path}/${id}`, { method: 'PUT', body, entityId })

export type OwnerBody = { code: string; name: string; note?: string; is_active?: boolean }
export type ProductBody = {
  code: string
  name: string
  unit: string
  barcode?: string
  category?: string
  owner_id?: string
  sale_price_idr: IDR
  is_active?: boolean
}
export type SupplierBody = {
  code: string
  name: string
  npwp?: string
  address?: string
  phone?: string
  issues_faktur: boolean
  is_active?: boolean
}
export type CustomerBody = {
  code: string
  name: string
  npwp?: string
  nik?: string
  address?: string
  phone?: string
  is_active?: boolean
}

export const listOwners = scoped<Owner>('/owners')
export const createOwner = creator<Owner, OwnerBody>('/owners')
export const updateOwner = updater<Owner, OwnerBody>('/owners')

export const listProducts = scoped<Product>('/products')
export const createProduct = creator<Product, ProductBody>('/products')
export const updateProduct = updater<Product, ProductBody>('/products')

export const listSuppliers = scoped<Supplier>('/suppliers')
export const createSupplier = creator<Supplier, SupplierBody>('/suppliers')
export const updateSupplier = updater<Supplier, SupplierBody>('/suppliers')

export const listCustomers = scoped<Customer>('/customers')
export const createCustomer = creator<Customer, CustomerBody>('/customers')
export const updateCustomer = updater<Customer, CustomerBody>('/customers')
